src/cache/algorithms/expr_variable_reordering.h

```diff
@@ -0,0 +1,189 @@
+   *
+   * @param expr relation
+   */
+  void run_on_relation(expr2tc &expr) noexcept;
+
+  /**
+   * Parses and adds the symbol or value
+   *
+   * This receives a reference and generates a copy of it to be later used
+   *
+   * @param op parent of the operation
+   * @param is_lhs truth value to determine which side of the op will be checked
+   * @param symbols
+   * @param values
+   */
+  void add_value(
+    const std::shared_ptr<arith_2ops> op,
+    bool is_lhs,
+    symbols_vec &symbols,
+    values_vec &values);
+
+  /**
+   * Parses and replaces symbols and values
+   *
+   * This works by tacking an expression and manually replacing it's inner
+   * expressions with the ordered symbols and values
+   *
+   * Example:
+   *
+   * op: (ADD (b) ( (ADD (7) (a) )
+   * symbols: ["a", "b"]
+   * values: [7]
+   *
+   * The expression would be parsed/replaced in the following order:
+   *
+   * (ADD (1) ( (ADD (2) (3)) )
+   *
+   * 1 -> "a"
+   * 2 -> "b"
+   * 3 -> "7"
+   *
+   * @param op parent of the operation
+   * @param is_lhs truth value to determine which side of the op will be checked
+   * @param symbols
+   * @param values
+   */
+  void replace_value(
+    const std::shared_ptr<arith_2ops> op,
+    bool is_lhs,
+    symbols_vec &symbols,
+    values_vec &values);
+
+  enum class PARSE_AS
+  {
+    BIN_OP,   /// For binary operations
+    CONSTANT, /// for constants
+    RELATION, /// for relations
+    SYMBOL,   /// for symbols
+    SKIP      /// for expressions that shouldn't be analyzed
+  };
+
+  enum class TRANSVERSE_MODE
+  {
+    READ,    /// to extract the expressions
+    REPLACE, /// to replace the expressions
+  };
+  /**
+   * Check expression and categorize it's type
+   * @param expr to be parsed
+   * @return PARSE_AS type of expression
+   */
+  static PARSE_AS get_expr_type(expr2tc &expr);
+
+  /**
+   * Default strategy to execute an algorithm while transversing the expr
+   * @param op binary operation
+   * @param symbols vector of all symbols expressions
+   * @param values vector of all values expressions
+   */
+  void transverse_binop(
+    const std::shared_ptr<arith_2ops> op,
+    symbols_vec &symbols,
+    values_vec &values,
+    TRANSVERSE_MODE mode);
+
+  /**
+   * Recursively extracts all symbols and values of a binary operation
+   *
+   * The logic is that an expr is like a lisp:
+   *
+   *  (BINOP (BINOP (x) (y)) (7))
+   *
+   *  So the output would be:
+   *
+   *  symbols = [x,y], values = [7]
+   *
+   *  TODO: Currently there is no logic to handle different BINOP inside the same expression
+   *
+   * @param op binary operation
+   * @param symbols vector of all symbols expressions
+   * @param values vector of all values expressions
+   */
+  void transverse_read_binop(
+    const std::shared_ptr<arith_2ops> op,
+    symbols_vec &symbols,
+    values_vec &values);
+
+  /**
+   * After transverse reading the all symbols and value this will
```

Comment:
This sentence is broken.

Reply:
Changed to "Transverse the expression replacing its inner symbols and values in alphabetical order"